File: app/src/main/java/Controller/WishListRecyclerViewAdapter.java
package Controller;

import android.content.Context;
import android.graphics.BitmapFactory;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.QuickContactBadge;
import android.widget.TextView;

import androidx.annotation.IntegerRes;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.elsafa.R;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;

import java.util.List;
import java.util.Map;

import Model.Item;
import Model.Product.Product;

public class WishListRecyclerViewAdapter extends RecyclerView.Adapter<WishListRecyclerViewAdapter.ViewHolder> {

    private FirebaseFirestore fStore;
    private List<Item> list;
    private Context context;
    private FirebaseAuth auth;


    public WishListRecyclerViewAdapter(List<Item> list, Context context) {
        this.list = list;
        this.context = context;
        fStore = FirebaseFirestore.getInstance();
        auth = FirebaseAuth.getInstance();
    }

    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.wishlist_item_card_view, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        fStore.collection("Categories")
                .document(list.get(position).getCategoryName())
                .collection("Products")
                .document(list.get(position).getProductId())
                .get()
                .addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
                    @Override
                    public void onSuccess(DocumentSnapshot ds) {
                        Product product = ds.toObject(Product.class);
                        product.setProductId(ds.getId());

                        holder.itemImage.setImageBitmap(BitmapFactory.
                                decodeByteArray(product.getMainPic().toBytes(), 0, product.getMainPic().toBytes().length));
                        holder.title.setText(product.getTitle());

                        int quantity = list.get(position).getQuantity();
                        float totalPrice = (product.getPrice() * quantity);
                        holder.price.setText(String.valueOf(totalPrice) + " EGP");
                    }
                });

        holder.remove.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                removeWishFromProduct(position);
            }
        });


        holder.addToCart.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                addCartToProduct(position);
            }
        });
    }

    private void removeWishFromProduct(int position) {
        fStore.collection("Categories")
                .document(list.get(position).getCategoryName())
                .collection("Products")
                .document(list.get(position).getProductId())
                .get()
                .addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
                    @Override
                    public void onSuccess(DocumentSnapshot ds) {
                        Map<String, Integer> map = (Map<String, Integer>) ds.get("wishes");
                        map.remove(auth.getUid());
                        fStore.collection("Categories")
                                .document(list.get(position).getCategoryName())
                                .collection("Products")
                                .document(list.get(position).getProductId())
                                .update("wishes", map);
                        removeProductFromWishList(position);
                    }
                });
    }

    private void addCartToProduct(int position) {
        fStore.collection("Categories")
                .document(list.get(position).getCategoryName())
                .collection("Products")
                .document(list.get(position).getProductId())
                .get()
                .addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
                    @Override
                    public void onSuccess(DocumentSnapshot ds) {
                        Map<String, Integer> map = (Map<String, Integer>) ds.get("carts");
                        map.put(auth.getUid(), 1);
                        fStore.collection("Categories")
                                .document(list.get(position).getCategoryName())
                                .collection("Products")
                                .document(list.get(position).getProductId())
                                .update("carts", map);

                        addProductToCart(position);
                    }
                });
    }

    private void addProductToCart(int position) {
        fStore.collection("Users")
                .document(auth.getUid())
                .collection("Cart")
                .document(list.get(position).getCategoryName() + list.get(position).getProductId())
                .set(list.get(position));

        removeWishFromProduct(position);
    }

    private void removeProductFromWishList(int position) {
        fStore.collection("Users")
                .document(auth.getUid())
                .collection("WishList")
                .document(list.get(position).getCategoryName() + list.get(position).getProductId())
                .delete();

        list.remove(position);
        notifyDataSetChanged();
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        ImageView itemImage;
        TextView title, price, remove, addToCart;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            itemImage = itemView.findViewById(R.id.item_image);
            title = itemView.findViewById(R.id.title);
            price = itemView.findViewById(R.id.price);
            remove = itemView.findViewById(R.id.remove);
            addToCart = itemView.findViewById(R.id.add_to_cart);
        }
    }
}

File: app/src/main/java/com/whiteside/cafe/adapter/WishListRecyclerViewAdapter.kt
package com.whiteside.cafe.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.whiteside.cafe.common.BaseListener
import com.whiteside.cafe.common.BaseViewHolder
import com.whiteside.cafe.databinding.WishlistItemCardViewBinding
import com.whiteside.cafe.model.Item
import com.whiteside.cafe.model.Product
import com.whiteside.cafe.ui.cart.CartPresenter
import com.whiteside.cafe.ui.product.ProductPresenter
import com.whiteside.cafe.ui.wishlist.WishListPresenter
import com.whiteside.cafe.utils.BlobBitmap
import javax.inject.Inject

class WishListRecyclerViewAdapter @Inject constructor(
    var productPresenter: ProductPresenter,
    var cartPresenter: CartPresenter,
    var wishListPresenter: WishListPresenter
) : RecyclerView.Adapter<WishListRecyclerViewAdapter.ViewHolder?>() {

    var list: ArrayList<Item> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        return ViewHolder(
            WishlistItemCardViewBinding.inflate(LayoutInflater.from(parent.context))
        )
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        productPresenter.getProduct(list[position].categoryName, list[position].productId) {
            holder.updateView(it, list[position])
        }
    }

    override fun getItemCount(): Int {
        return list.size
    }

    inner class ViewHolder(val bind: WishlistItemCardViewBinding) :
        BaseViewHolder(bind) {

        fun updateView(product: Product, item: Item) {
            bind.itemImage.setImageBitmap(BlobBitmap.convertBlobToBitmap(product.productPics[0]))
            bind.title.text = product.title
            bind.price.text = "${product.price * item.quantity!!}"

            bind.remove.setOnClickListener {
                wishListPresenter.removeItem(product,
                    object : BaseListener<Product> {
                        override fun onSuccess(result: Product) {
                            handleSuccess()

                            bind.root.removeAllViews()
                            Toast.makeText(
                                bind.root.context,
                                "Item removed from wishlist",
                                Toast.LENGTH_SHORT
                            ).show()
                        }

                        override fun onFailed(exception: Exception) {
                            handleFailure(exception)
                        }

                        override fun onLoading() {
                            handleLoading()
                        }
                    })
            }

            bind.addToCart.setOnClickListener {
                cartPresenter.addItem(product,
                    object : BaseListener<Product> {
                        override fun onSuccess(result: Product) {
                        }

                        override fun onFailed(exception: Exception) {
                            handleFailure(exception)
                        }

                        override fun onLoading() {
                        }
                    })

                wishListPresenter.removeItem(product,
                    object : BaseListener<Product> {
                        override fun onSuccess(result: Product) {
                            handleSuccess()

                            bind.root.removeAllViews()
                            Toast.makeText(
                                bind.root.context,
                                "Item added from wishlist",
                                Toast.LENGTH_SHORT
                            ).show()
                        }

                        override fun onFailed(exception: Exception) {
                            handleFailure(exception)
                        }

                        override fun onLoading() {
                            handleLoading()
                        }
                    })
            }
        }
    }
}
